src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                                {
+                                    response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
+                                }
+                            }
+                        }
+
+                        return size;
+                    }
+                    catch (Exception ex)
+                    {
+                        easy.FailRequest(ex); // cleanup will be handled by main processing loop
+                    }
+                }
+
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error
+                return size - 1;
+            }
+
+            private unsafe static size_t CurlReceiveBodyCallback(
+                byte* buffer, size_t size, size_t nitems, IntPtr context)
+            {
+                size *= nitems;
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
+                {
+                    try
+                    {
+                        if (!(easy.Task.IsCanceled || easy.Task.IsFaulted))
+                        {
+                            // Complete the task if it hasn't already been.  This will make the
+                            // stream available to consumers.  A previous write callback
+                            // may have already completed the task to publish the response.
+                            easy.EnsureResponseMessagePublished();
+
+                            // Wait for a reader
+                            // TODO: The below call blocks till all the data has been read since
+                            //       response body is not suppored to be buffered in memory.
+                            //       Figure out some way to work around this
+                            if (size != 0)
+                            {
+                                easy.ResponseMessage.ContentStream.WaitAndSignalReaders(buffer, (long)size);
+                            }
+
+                            return size;
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        easy.FailRequest(ex); // cleanup will be handled by main processing loop
+                    }
+                }
+
+                // Returing a value other than size fails the callback and forces
+                // request completion with an error.
+                return (size > 0) ? size - 1 : 1;
+            }
+
+            private static size_t CurlSendCallback(IntPtr buffer, size_t size, size_t nitems, IntPtr context)
+            {
+                size *= nitems;
+                if (size == 0)
+                {
+                    return 0;
+                }
+
+                EasyRequest easy;
+                if (TryGetEasyRequestFromContext(context, out easy))
```

Comment:
We could, but I'm not convinced it's necessary.  We're now registering for cancellation notifications via CancellationToken.Register.  If cancellation is requested, the curl_multi_wait will wake up, loop around, and as part of processing requests will see that cancellation was requested and will do all of the appropriate cleanup.  Is there a case I'm missing?  Is the concern that this will end up doing lots of work before we loop around again?